Only call material computeProperties when computing thread actually needs them

lindsayad opened the pull request:
This material property dependency behaves analogously to variable dependencies with similar methods in `MooseObjectWarehouseBase`, `MaterialPropertyInterface` vs `MooseVariableDependencyInterface`, and in the computing threads: residual, jacobian, aux_vars, etc.

One test case submitted that ensures that the elem\_aux\_kernel thread doesn't call material compute properties.

Couple of notes:

1. This PR doesn't currently help if there are multiple materials defined on a block, some providing properties and others not; as long as one material is supplying a property on a block, all the materials `computeQpProperties` methods will get called
2. Along the same lines, the set of active material property ids aren't currently being used for any purpose. More work could be done to do an inverse mapping of material properties needed to materials needed, and then item 1 above could be addressed.

Closes  #8444

lindsayad committed:
Initial work on mat prop dependencies (#8444).

lindsayad committed:
Add getPropId method to MaterialData.

lindsayad committed:
Change set from containing MaterialProperty pointers to unsigned ints representing mat prop ids (#8444).

lindsayad committed:
Update material portions (#8444).

lindsayad committed:
Add new checks and methods in FEProblemBase and SubProblem (#8444).

lindsayad committed:
Add test showing that new check FEProblemBase::reinitMaterials has effect (#8444).

lindsayad committed:
Move updateMatPropDependency into MooseObjectWarehouseBase. (#8444)

- Also make material computing threads (e.g. ComputeResidualThread call the
  updateMatPropDependency methods)
- Make sure that we don't mark as having material property dependencies simply
  because we call setActiveMaterialProperties in FEProblemBase/Subproblem
- Move from MooseSharedPointer to std::shared_ptr
- Closed space between right angle brackets

lindsayad committed:
Make sure to clearActiveMaterialProperties after compute threads (#8444)

- Test 1: don't call computeQpProperties for one material that doesn't supply
  any properties - Passes
- Test 2: one material supplies properties. The other does not. Ideally only the
  one supplying properties should get called - Fails
- Test 3: ensure that materials get computed for residual and jacobian threads
  because kernel needs property but that material does not get computed for
  elem_aux_var that does not need any properties

lindsayad committed:
Add IncrementMaterial test (#8444).

lindsayad committed:
Add dependencies for other getProperty calls (#8444).

lindsayad committed:
Fix compute thread for ElemAuxBcs (#8444).

lindsayad committed:
Fix dirac kernels thread and LineMaterialSamplerBase thread. (#8444).

lindsayad committed:
Reenable resetQpProperties warning. (#8444).

lindsayad committed:
Fix up ComputeElemAuxBcsThread and regold random_material. (#8444)

andrsd commented on framework/include/base/MooseObjectWarehouseBase.h:
```diff
@@ -160,6 +168,9 @@ class MooseObjectWarehouseBase
                                              const std::vector<MooseSharedPointer<T> > & objects);
 
+  static void updateMatPropDependencyHelper(std::set<unsigned int> & needed_mat_props,
+                                             const std::vector<MooseSharedPointer<T> > & objects);
```
Align the `const` with `std:...`, i.e. remove 1 space on this line ;-)

andrsd commented on framework/include/base/MooseObjectWarehouseBase.h:
```diff
@@ -539,6 +550,56 @@ MooseObjectWarehouseBase<T>::updateVariableDependencyHelper(std::set<MooseVariab
 
+template<typename T>
+void
+MooseObjectWarehouseBase<T>::updateMatPropDependency(std::set<unsigned int> & needed_mat_props, THREAD_ID tid /* = 0*/) const
```
An extra space between `tid` and `/*`. The same in `updateBlockMatPropDependency` below.

andrsd commented on framework/include/base/SubProblem.h:
```diff
@@ -111,6 +111,14 @@ class SubProblem : public Problem
 
+  virtual void setActiveMaterialProperties(const std::set<unsigned int> & mat_prop_ids, THREAD_ID tid);
+
+  virtual const std::set<unsigned int> &getActiveMaterialProperties(THREAD_ID tid);
```
Add a space between `&` and the function name.

andrsd commented on framework/include/materials/MaterialData.h:
```diff
@@ -129,6 +129,8 @@ class MaterialData
    */
   const MaterialPropertyStorage & getMaterialPropertyStorage() const { return _storage; }
 
+  unsigned int getPropertyId (const std::string & prop_name) const { return _storage.getPropertyId(prop_name); }
```
No space between the method name and `(`.

andrsd commented on test/include/materials/IncrementMaterial.h:
```diff
@@ -0,0 +1,38 @@
+InputParameters validParams<GenericConstantMaterial>();
+
+/**
+ * GenericConstantMaterial that tracks the number of times computeQpProperties has been called.
```
Just say "A material that tracks..."

lindsayad commented:
I need to fix the new test to address threading.

lindsayad committed:
No threading allowed for increment test (#8444).

YaqiWang commented:
Thanks @lindsayad This PR is really important to us.

lindsayad commented:
I can't see the app tests, so I don't know why they're failing. One possibility is if they have `GeneralUserObjects` that call `FEProblemBase::reinitMaterials`

YaqiWang commented:
Yak does. Any suggestion to fix it?

lindsayad commented:
You can see an example of what needs to be done in `LineMaterialSamplerBase.h` from lines 150-153, and then line 187. So before the element loop:
```
  std::set<unsigned int> needed_mat_props;
  const std::set<unsigned int> & mp_deps = this->getMatPropDependencies();
  needed_mat_props.insert(mp_deps.begin(), mp_deps.end());
  _fe_problem.setActiveMaterialProperties(needed_mat_props, _tid);
```
and then after the element loop: `_fe_problem.clearActiveMaterialProperties(_tid);`

If there are other test failures not related to `GeneralUserObjects` I'll need some help figuring them out.

permcody commented:
It looks like 3 tests in BISON are failing (Exodiff) and 4 tests in
Mastodon (Exodiff). I think we'll have to dive in to figure out why on our
side aside from Yaqi's tests. It might be more of the same, other than that
this is looking really good!

On Mon, Feb 6, 2017 at 1:12 PM Alex Lindsay <notifications@github.com>
wrote:

> You can see an example of what needs to be done in
> LineMaterialSamplerBase.h from lines 150-153, and then line 187. So
> before the element loop:
>
>   std::set<unsigned int> needed_mat_props;
>   const std::set<unsigned int> & mp_deps = this->getMatPropDependencies();
>   needed_mat_props.insert(mp_deps.begin(), mp_deps.end());
>   _fe_problem.setActiveMaterialProperties(needed_mat_props, _tid);
>
> and then after the element loop:
> _fe_problem.clearActiveMaterialProperties(_tid);
>
> If there are other test failures not related to GeneralUserObjects I'll
> need some help figuring them out.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/8467#issuecomment-277798982>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AC5XIDAwCI7ByxQQmsleqfShUF5rcok8ks5rZ36OgaJpZM4L1wRM>
> .
>

YaqiWang commented:
Yes, my feeling is that your guys can start reviewing process. We possibly can patch Yak later. I am in middle of something, @snschune since those are your codes, can you find some test to try Alex's suggestions? I might get time later tomorrow if Sebastian is busy.

snschune commented:
> One possibility is if they have GeneralUserObjects that call FEProblemBase::reinitMaterials

Could be the case. All NDA DFEM tests fail, I'll take a look.

permcody commented on framework/include/base/FEProblemBase.h:
```diff
@@ -252,6 +252,14 @@ class FEProblemBase :
    */
   virtual void clearActiveElementalMooseVariables(THREAD_ID tid) override;
 
+  virtual void setActiveMaterialProperties(const std::set<unsigned int> & mat_prop_ids, THREAD_ID tid) override;
```
Would you mind putting in comments for each method you added?

permcody commented on framework/include/base/MooseObjectWarehouseBase.h:
```diff
@@ -539,6 +550,56 @@ MooseObjectWarehouseBase<T>::updateVariableDependencyHelper(std::set<MooseVariab
+{
+  if (hasActiveBoundaryObjects(tid))
+  {
+    typename std::map<BoundaryID, std::vector<MooseSharedPointer<T> > >::const_iterator it;
```
This is OK and I know it's copying from other code but we can just use `auto` now for these long declarations. In fact, you can just use a range based for loop here to make this really clean and tight!

permcody commented on framework/include/base/MooseObjectWarehouseBase.h:
```diff
@@ -539,6 +550,56 @@ MooseObjectWarehouseBase<T>::updateVariableDependencyHelper(std::set<MooseVariab
+void
+MooseObjectWarehouseBase<T>::updateMatPropDependencyHelper(std::set<unsigned int> & needed_mat_props, const std::vector<MooseSharedPointer<T> > & objects)
+{
+  for (typename std::vector<MooseSharedPointer<T> > ::const_iterator it = objects.begin(); it != objects.end(); ++it)
```
`auto` should work here and in other places (I think). Never tried it in a place where typename is required.

permcody commented on framework/include/base/SubProblem.h:
```diff
@@ -111,6 +111,14 @@ class SubProblem : public Problem
    */
   virtual void clearActiveElementalMooseVariables(THREAD_ID tid);
 
+  virtual void setActiveMaterialProperties(const std::set<unsigned int> & mat_prop_ids, THREAD_ID tid);
```
More method descriptions please.

permcody commented on framework/include/base/SubProblem.h:
```diff
@@ -368,6 +376,10 @@ class SubProblem : public Problem
   /* This needs to remain <unsigned int> for threading purposes */
   std::vector<unsigned int> _has_active_elemental_moose_variables;
 
+  std::vector<std::set<unsigned int> > _active_material_property_ids;
```
We are starting to close the space in `> >` now. Feel free to do that in new code you commit.

permcody reviewed (commented):
Looking good. Please address these comments and then we can take it from here to get the other apps patched as needed. Great work!

lindsayad committed:
Add method documentation (#8444)

lindsayad committed:
Convert to range based for loops (#8444)

lindsayad commented:
I think it's interesting that the Intel tester doesn't know how to spell: `You disabeed the computatoon of this (mat) material by MOOSE`

permcody commented:
You are seeing corruption of the output stream due to a thread race. We are working on changing mooseError and mooseWarning and will include a thread mutex to prevent this from happening.

snschune commented:
> You are seeing corruption of the output stream due to a thread race. We are working on changing mooseError and mooseWarning and will include a thread mutex to prevent this from happening.

With mpi, I've had cases where `mooseError` was issued on `process != 0` and it got swallowed. Are the changes going to address that, too? Of course you can run with `--keep-cout` but then the output is a huge mess.

permcody commented:
@snschune - This discussion is diverging from this ticket. We might want to talk about the changes over here: #8411. BTW - You'd have to show me this issue with mooseError(), that should *never* get swallowed. However, it's possible to lose information in the console buffer that wasn't flushed so you lose output but not the error. 

We are well aware of this and will deal with it. I'm not sure if it'll be in #8411 or not but we are working on it.

permcody commented:
@idaholab/moose-team - @lindsayad is finished with as much as he can contributed on this PR but we need to help him get it merged. It's failing a few BISON tests (probably due to legacy problems) but other than that it looks really good. It has the potential to speed up a lot of simulations. Any volunteers?

rwcarlsen commented on framework/src/base/FEProblemBase.C:
```diff
@@ -2016,25 +2016,35 @@ void
 FEProblemBase::reinitMaterials(SubdomainID blk_id, THREAD_ID tid, bool swap_stateful)
 {
-  if (_all_materials.hasActiveBlockObjects(blk_id, tid))
+  if (_all_materials.hasActiveBlockObjects(blk_id, tid) && this->hasActiveMaterialProperties(tid))
```
Are both of these always true together?

lindsayad replied:
No. In your input file you could have a material defined on a block that is "active" but doesn't supply any properties that kernels, etc. need.

lindsayad replied:
However, it is possible that if the second check is true, then the second has to be true...

permcody replied:
Then these two tests can be reversed and the operator changed to an "or"

rwcarlsen replied:
That's what I was thinking - @permcody - does that seem to be right - the total expression should always be true if the second part is true?

rwcarlsen replied:
If we change anything, it should probably be switched in all similar expressions in this PR.

permcody replied:
I don't see how the second expression could be true without the first, yes.

lindsayad replied:
I can see the logic in switching the order for when `hasActiveMaterialProperties` evaluates to `false` and we use "and", but I don't see how the "or" would be a good idea. Because it would defeat the whole purpose of the PR if `hasActiveMaterialProperties` evaluated to 'false` and `hasActiveBlockObjects` evaluated to `true` and we used an "or" comparison.

rwcarlsen commented on framework/include/base/SubProblem.h:
```diff
@@ -368,6 +400,12 @@ class SubProblem : public Problem
+  std::vector<std::set<unsigned int>> _active_material_property_ids;
+
+  /// True if _active_material_property_ids is not empty
+  std::vector<unsigned int> _has_active_material_property_ids;
```
This should be ``vector<bool>`` and all usages should use ``true`` and ``false``.  It might actually be worth just using ``_active_material_property_ids[thread_id].empty()`` instead and removing this second vector entirely - then we don't have to worry about the two getting out of sync.

permcody replied:
There are certain members of the team that are terrified and completely against vector<bool> because it has bitten them (us) before in threaded cases. You won't find it much in the framework because of that. I'm not part of that group but it's just not worth fighting for the most part.

He's following existing patterns here. We can clean them all up later.

rwcarlsen replied:
Easy and clean to just remove the second vector entirely though.  I'm in favor of doing it well the first time - we've all experienced the longevity of temporary solutions ;)

permcody replied:
Oh I see, I was too busy worry about the `vector<bool>` comment. Yes, I believe we could drop the second vector here without any issue.

friedmud replied:
> There are certain members of the team that are terrified and completely against vector because it has bitten them (us) before in threaded cases. 

You say this like my dislike of `std::vector<bool>` is unjustified... but it's not.  Indeed, in this very case: if you were to use a `std::vector<bool>` you would have a thread race condition that would lead to incorrect execution!

There are very real and VERY good reasons for avoiding `std::vector<bool>` in a code that relies on threading.  There is absolutely no reason to use it until you can justify that the memory savings is worth it... and then that variable should have a ton of comments around it telling people what's up.

permcody replied:
Indeed - This *is* the case where `vector<bool>` will screw you. 😄

rwcarlsen commented on framework/include/materials/MaterialPropertyInterface.h:
```diff
@@ -234,6 +242,18 @@ class MaterialPropertyInterface
+   * @param mat_prop_id The material property ID that should be added to _this_ object's material
+   * dependency list.
+   */
+  void addMatPropDependency(unsigned int mat_prop_id) { _material_property_dependencies.insert(mat_prop_id); }
```
I don't think we need a protected function.  It is only used in this class and not meant to be overridden.  Better to just do the one line insert directly in each location and not have this function altogether.

rwcarlsen commented on framework/include/materials/MaterialData.h:
```diff
@@ -129,6 +129,15 @@ class MaterialData
+   *
+   * @return An unsigned int corresponding to the property ID of the passed in prop_name
+   */
+  unsigned int getPropertyId(const std::string & prop_name) const { return _storage.getPropertyId(prop_name); }
```
We don't really need this function I think.  We can just use ``  MaterialData::getMaterialPropertyStorage().getPropertyId(prop_name)``.

friedmud replied:
I would rather have this function.  It's logical to be able to ask this class for the ID... and it's inconsequential how the material property IDs are being held internally.

rwcarlsen replied:
It's too bad the interface already leaks the underlying storage object at all. :-/

rwcarlsen commented on framework/include/vectorpostprocessors/LineMaterialSamplerBase.h:
```diff
@@ -147,6 +147,11 @@ LineMaterialSamplerBase<T>::execute()
   std::vector<Real> values(_material_properties.size());
 
+  std::set<unsigned int> needed_mat_props;
+  const std::set<unsigned int> & mp_deps = this->getMatPropDependencies();
```
``this`` is implied inside objects and we generally don't use it.

rwcarlsen commented on framework/include/vectorpostprocessors/LineMaterialSamplerBase.h:
```diff
@@ -179,6 +184,7 @@ LineMaterialSamplerBase<T>::execute()
       addSample(_q_point[qp], qp_proj_dist_along_line, values);
     }
   }
+  _fe_problem.clearActiveMaterialProperties(_tid);
```
There is a lot of this pattern (e.g in all the compute threads classes) - setting up active material properties state and then clearing it with important code in between.  I'd be worried about this being difficult to keep track of - easy for bugs to appear.  Someone might skip the clear with an early return for example.  Or someone is surprised when there material properties aren't being recomputed as expected because they didn't know they needed to go through this explicit property id adding process.

* Could we somehow make this safer so that people don't get surprised by that?
* Is there a better place we could put the clear to be more "automatic" - and possible a place that would remove the need for certain classes to have to re-add active properties temporarily like here?

permcody replied:
We can always use RAII, but as you pointed out he's just following a pattern here. All of this code/logic looks like the variable dependency stuff. It's not really a hold up for this PR. RAII can have runtime costs so we have to consider those tradeoffs.

rwcarlsen replied:
The material property dependency setup part is a new pattern though - one that seems like it will be easy to get wrong and bite people.

rwcarlsen replied:
The setup and teardown in the compute threads classes occurs once per looping over the mesh - not a problem.  But the one here, for example causes the prop dependency setup/teardown to occur for every element this postprocessor executes on.

friedmud replied:
Does it?  It looks like this happens after the element loop to me.

permcody replied:
Right this is **outside** the elemental loop. This is a `GeneralUserObject` which is why it doesn't automatically know about material dependencies. These cases are rare but they aren't zero. Looking for a better solution is a bigger problem that will take some redesign and if we do come up with a solution, it'll go in several places (not just here).

rwcarlsen replied:
You are both correct.  The computational cost issue has been refuted, but I'd like to see if we can think of a better way for object not within these automatically handled thread loop cases to at least know they need to init material properties.  Maybe something like a warning if they try to compute material properties but zero/none are actually computed?

friedmud replied:
Can't warn in that case... that happens all the time.

permcody replied:
Yeah - you only get screwed when you make your own loop. All framework loops are handled automatically. It's really hard to warn unless we take away people's ability to make their own loops. That's probably not going to happen.

friedmud replied:
I mean: we could make an `ElementLoopingGeneralUserObject` base class that works similarly to our `ThreadedElementLoop` with virtual function callbacks users can fill in and there would be some ability to feed an element list down to the base-class as well.

Just depends on how many times this is being done in the applications... if it's only a couple then it's not worth a base-class.  If it's more than 4 or 5 I would say we should abstract it out so people aren't doing it themselves.

rwcarlsen replied:
We have one case in the framework.  A good look at the failing tests will probably be able to answer the rest of that question.

rwcarlsen reviewed (changes requested):
Looks pretty good.  There are a few things I think are worth thinking a bit about before we commit - i.e. how can we make this "easy" for users/app-developers to use correctly and not be surprised.  Inline comments have more detail.

rwcarlsen commented:
If we change anything, it should probably be switched in all similar
expressions in this PR.

On Thu, Feb 9, 2017 at 10:15 AM, Cody Permann <notifications@github.com>
wrote:

> *@permcody* commented on this pull request.
> ------------------------------
>
> In framework/src/base/FEProblemBase.C
> <https://github.com/idaholab/moose/pull/8467>:
>
> >  }
>
>  void
>  FEProblemBase::reinitMaterials(SubdomainID blk_id, THREAD_ID tid, bool swap_stateful)
>  {
> -  if (_all_materials.hasActiveBlockObjects(blk_id, tid))
> +  if (_all_materials.hasActiveBlockObjects(blk_id, tid) && this->hasActiveMaterialProperties(tid))
>
> Then these two tests can be reversed and the operator changed to an "or"
>
> —
> You are receiving this because you are on a team that was mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/8467>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABSeSjhk_xvSkjAXc0YvzX_Lxl2oP7mQks5ra0nMgaJpZM4L1wRM>
> .
>

lindsayad committed:
Address rwcarlsen review (#8444).

- remove _has_active_material_property_ids from SubProblem.h
- remove addMatPropDependency from MaterialPropertyInterface.h
- remove this-> calls because it's implied you dope
- swap checks in FEProblemBase

permcody commented:
This is based on the assumption that you gave earlier that
"hasActiveMaterialProperties" is always true if "hasActiveBlockObjects".
However looking at this again, I pretty sure that assumption is not true.
Switching won't do anything for us!

On Thu, Feb 9, 2017 at 6:53 PM Alex Lindsay <notifications@github.com>
wrote:

> *@lindsayad* commented on this pull request.
> ------------------------------
>
> In framework/src/base/FEProblemBase.C
> <https://github.com/idaholab/moose/pull/8467>:
>
> >  }
>
>  void
>  FEProblemBase::reinitMaterials(SubdomainID blk_id, THREAD_ID tid, bool swap_stateful)
>  {
> -  if (_all_materials.hasActiveBlockObjects(blk_id, tid))
> +  if (_all_materials.hasActiveBlockObjects(blk_id, tid) && this->hasActiveMaterialProperties(tid))
>
> I can see the logic in switching the order for when
> hasActiveMaterialProperties evaluates to false and we use "and", but I
> don't see how the "or" would be a good idea. Because it would defeat the
> whole purpose of the PR if hasActiveMaterialProperties evaluated to 'false
> andhasActiveBlockObjectsevaluated totrue` and we used an "or" comparison.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/8467>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AC5XIG45fF-9TGtLr-mkEMx-cKMpqAjuks5ra8MigaJpZM4L1wRM>
> .
>

lindsayad commented:
I was trying to say it the other way around: if "hasActiveMaterialProperties" is true, then "hasActiveBlockObjects" must be true or else the material checks that happen at the start of a simulation have let us down (kind of like in #5309)

rwcarlsen commented:
Should we try removing the hasActiveBlockObjects condition then and see if
it breaks anything?

On Fri, Feb 10, 2017 at 8:46 AM, Alex Lindsay <notifications@github.com>
wrote:

> I was trying to say it the other way around: if
> "hasActiveMaterialProperties" is true, then "hasActiveBlockObjects" must be
> true or else the material checks that happen at the start of a simulation
> have let us down (kind of like in #5309
> <https://github.com/idaholab/moose/issues/5309>)
>
> —
> You are receiving this because you are on a team that was mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/8467#issuecomment-278979559>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABSeSosIegOMRDjDW3hsuuhVXyPKz2u8ks5rbIZqgaJpZM4L1wRM>
> .
>

lindsayad commented:
Sure, easy enough to try!

On 02/10/2017 10:25 AM, Robert Carlsen wrote:
> Should we try removing the hasActiveBlockObjects condition then and see if
> it breaks anything?
>
> On Fri, Feb 10, 2017 at 8:46 AM, Alex Lindsay <notifications@github.com>
> wrote:
>
> > I was trying to say it the other way around: if
> > "hasActiveMaterialProperties" is true, then "hasActiveBlockObjects" 
> must be
> > true or else the material checks that happen at the start of a 
> simulation
> > have let us down (kind of like in #5309
> > <https://github.com/idaholab/moose/issues/5309>)
> >
> > —
> > You are receiving this because you are on a team that was mentioned.
> > Reply to this email directly, view it on GitHub
> > 
> <https://github.com/idaholab/moose/pull/8467#issuecomment-278979559>, 
> or mute
> > the thread
> > 
> <https://github.com/notifications/unsubscribe-auth/ABSeSosIegOMRDjDW3hsuuhVXyPKz2u8ks5rbIZqgaJpZM4L1wRM>
> > .
> >
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://github.com/idaholab/moose/pull/8467#issuecomment-278990165>, 
> or mute the thread 
> <https://github.com/notifications/unsubscribe-auth/AJxgcAV1uPWxtYVAG0VWCrBTAEpFL0CAks5rbI-QgaJpZM4L1wRM>.
>

lindsayad committed:
Try removing hasActiveBlocks check (#8444)

lindsayad commented:
I guess we're just mulling over the App tests now?

rwcarlsen commented:
The if clause change didn't seem to break anything new.  @lindsayad - most of the rest we'll try to shake out on our end with the app tests - we'll ping if we need you again.

permcody commented:
@lindsayad - Haven't forgotten about this PR. Will get to it soon. Yak has several failing tests AND we are planning on merging it with another application so we have a few more steps before we can merge this in.

lindsayad commented:
No worries. Thanks!

On 02/15/2017 11:36 AM, Cody Permann wrote:
>
> @lindsayad <https://github.com/lindsayad> - Haven't forgotten about 
> this PR. Will get to it soon. Yak has several failing tests AND we are 
> planning on merging it with another application so we have a few more 
> steps before we can merge this in.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://github.com/idaholab/moose/pull/8467#issuecomment-280080878>, 
> or mute the thread 
> <https://github.com/notifications/unsubscribe-auth/AJxgcASZk4nCVmUnndWZ2h7hcenZBRLuks5rczeMgaJpZM4L1wRM>.
>

rwcarlsen committed:
cleanup dirac

rwcarlsen committed:
make indicators inherit MaterialPropertyInterface and implement matprop deps for their thread

rwcarlsen commented:
I've been poking around the app test failures.
The failing BISON tests use the "output_properties" feature of material class - which creates an AuxKernel to track/output the value of the material property.  The ComputeUserObjects thread runs on initial and sim-end by default and so was (previously) running extra property evaluations at the beginning and end of the simulation which change the recorded values in the exodus file.  At least this is my current (possibly incorrect) understanding of the problem.

While poking around this, I had an idea.  It seems to me that most of the time, most material properties in any given MOOSE simulation will be used at least once during a given solve iteration.  Since this is the case, we could achieve most of the benefit of this kind of dependency resolution if we just remembered whether or not we had computed a particular property yet this iteration.  If we have already computed it, just have Material::computeProperties() do nothing.  Otherwise, it computes as normal.  Then provide some sort of material reset - that resets all the "_am_computed" material members to false.  This ensure at most one evaluation of each material per solve (instead of one evaluation per ``Compute..Thread``) and be more robust against people inheriting from MaterialPropertyInterface in their own objects where we miss dealing with the material property dependencies correctly with the current PR.

Worth looking into (either instead of or as a follow up to this PR)?

permcody commented:
>I've been poking around the app test failures. The failing BISON tests use the "output_properties" feature of material class - which creates an AuxKernel to track/output the value of the material property. The ComputeUserObjects thread runs on initial and sim-end by default and so was (previously) running extra property evaluations at the beginning and end of the simulation which change the recorded values in the exodus file. At least this is my current (possibly incorrect) understanding of the problem.

You've got it, we are computing and re-computing properties all the time, before, during and after.

>While poking around this, I had an idea. It seems to me that most of the time, most material properties in any given MOOSE simulation will be used at least once during a given solve iteration. Since this is the case, we could achieve most of the benefit of this kind of dependency resolution if we just remembered whether or not we had computed a particular property yet this iteration.

Well yes and no. The design of the material property system is that we don't store anything we compute. This has advantages and disadvantages of course. We only store stateful properties, everything else is computed on the fly and then immediately discarded. We recompute it again if there's another element loop. This design goes back to the beginning of MOOSE before we had 20+ pluggable systems. Back then we really did only need a single iteration through the mesh per iteration so there was no point in saving anything off. Stateful properties were added later but most properties don't require state.

> If we have already computed it, just have Material::computeProperties() do nothing. Otherwise, it computes as normal. Then provide some sort of material reset - that resets all the "_am_computed" material members to false. This ensures at most one evaluation of each material per solve (instead of one evaluation per Compute..Thread) and would be more robust against people inheriting from MaterialPropertyInterface in their own objects where we miss dealing with the material property dependencies correctly with the current PR.

This would be a huge change to the system. You'd have to store every property which can get expensive. It may be worth looking into in the future but it's trading time for space. There's no free lunch.

> Worth looking into (either instead of or as a follow up to this PR)?

Not right now. We know that material property storage can be expensive. Tensor mechanics requires the storage of a few tensors per quadrature point. This is one of those grand ideas we could look into in a bigger rewrite of the framework if that were to ever happen.

rwcarlsen commented:
I don't see how it would be a huge change.  It would be adding a member and a couple of functions to Material and FEProblem.  The properties are computed and stored - they only get clobbered if we call recompute them again on the same quad point, face, etc.  They are already there cached - pretty sure we just need to keep track of whether or not they've been computed.

rwcarlsen commented:
Never mind - @permcode seems to be right.

rwcarlsen commented:
Okay - after some more investigating, I have determined that non-idempotent materials in BISON are causing the failures.  Because this PR changes the number of times materials are evaluated - those tests break.  They will probably just need to be regolded :-/

rwcarlsen commented:
I've got a PR into rattlesnake that fixes the crashing tests.  Still need to investigate rattlesnake exodiff failures...

rwcarlsen committed:
compute thread housekeeping

rwcarlsen committed:
make materialdata safe when swap is called multiple times (e.g. via successive reinitMaterials calls without interleaving swapBackMaterials calls)

rwcarlsen commented:
I've spent all day looking at the rattlesnake exodiff failures and am still drawing a blank.  The *only* difference in material evaluations between the two runs (with and without this PR) is during ``FEProblem::initialSetup``'s call to set up user objects - with this PR, that call results in no material prop evaluations.  We can either toss the ball to rattlesnake team's court or patch initialSetup to fake/force evaluation of all materials during ComputeUserObjectsThread.

Also, I'm not totally confident in my previous reasoning for the BISON failures - should probably work with them a bit to confirm.

permcody commented:
Thanks for staying on this. @YaqiWang or @snschune perhaps you have some ideas on why Alex's patch might cause small diffs in a few tests?

snschune commented:
I'll take a look.

snschune commented:
@rwcarlsen can you give me access to your rattlesnake repo?

snschune commented:
As @rwcarlsen mentioned it's the fission source not being executed on initial that causes the issue. This appears only for cases where we set the initial condition of a transient via an eigenvalue.  
Adding a dummy pp like this
```
  [./a_dummy]
    type = ElementIntegralMaterialProperty
    mat_prop = fission_source
    execute_on = 'initial'
  [../]
```
fixes it [trivial diff about a_dummy being added]. 
We could add a PP in for transients via the ActionSystems @YaqiWang?
As far as I can see, only one test has this issue, the others should be iterative tolerances.

lindsayad commented:
It's for me to tell without seeing the Rattlesnake code, but is this 
truly some unorthodox calculation case? Because if the PR isn't allowing 
material calculations when it's supposed to be, then the PR isn't doing 
its job. However, if it's some heavily custom logic in the user object, 
then perhaps I can forgive it.

On 03/01/2017 10:31 PM, Sebastian Schunert wrote:
>
> As @rwcarlsen <https://github.com/rwcarlsen> mentioned it's the 
> fission source not being executed on initial that causes the issue. 
> This appears only for cases where we set the initial condition of a 
> transient via an eigenvalue.
> Adding a dummy pp like this
>
> |[./a_dummy] type = ElementIntegralMaterialProperty mat_prop = 
> fission_source execute_on = 'initial' [../] |
>
> fixes it [trivial diff about a_dummy being added].
> We could add a PP in for transients via the ActionSystems @YaqiWang 
> <https://github.com/YaqiWang>?
> As far as I can see, only one test has this issue, the others should 
> be iterative tolerances.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://github.com/idaholab/moose/pull/8467#issuecomment-283553372>, 
> or mute the thread 
> <https://github.com/notifications/unsubscribe-auth/AJxgcNQByC7GCqQCrGzaFz5QYPOZ-JQuks5rhkYIgaJpZM4L1wRM>.
>

permcody commented:
Alex, I don't think this is your issue. I believe Sebastian is saying
nothing is actually consuming that property on initial so your patch is
doing the correct thing. I don't like the dummy dependency solution but if
that's what Yaqi wants to do, that's fine.

On Thu, Mar 2, 2017 at 12:16 AM Alex Lindsay <notifications@github.com>
wrote:

> It's for me to tell without seeing the Rattlesnake code, but is this
> truly some unorthodox calculation case? Because if the PR isn't allowing
> material calculations when it's supposed to be, then the PR isn't doing
> its job. However, if it's some heavily custom logic in the user object,
> then perhaps I can forgive it.
>
> On 03/01/2017 10:31 PM, Sebastian Schunert wrote:
> >
> > As @rwcarlsen <https://github.com/rwcarlsen> mentioned it's the
> > fission source not being executed on initial that causes the issue.
> > This appears only for cases where we set the initial condition of a
> > transient via an eigenvalue.
> > Adding a dummy pp like this
> >
> > |[./a_dummy] type = ElementIntegralMaterialProperty mat_prop =
> > fission_source execute_on = 'initial' [../] |
> >
> > fixes it [trivial diff about a_dummy being added].
> > We could add a PP in for transients via the ActionSystems @YaqiWang
> > <https://github.com/YaqiWang>?
> > As far as I can see, only one test has this issue, the others should
> > be iterative tolerances.
> >
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/idaholab/moose/pull/8467#issuecomment-283553372>,
> > or mute the thread
> > <
> https://github.com/notifications/unsubscribe-auth/AJxgcNQByC7GCqQCrGzaFz5QYPOZ-JQuks5rhkYIgaJpZM4L1wRM
> >.
> >
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/8467#issuecomment-283558801>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AC5XIMhcsPyL2K5hN3RXvLEoNHIodro6ks5rhlCmgaJpZM4L1wRM>
> .
>

snschune commented:
I agree to Cody. Come to think about it again we should understand why the
evaluation on initial is important. And yes this is not the patch's issue,
it's Rattlesnake's.

On Mar 2, 2017 6:35 AM, "Cody Permann" <notifications@github.com> wrote:

> Alex, I don't think this is your issue. I believe Sebastian is saying
> nothing is actually consuming that property on initial so your patch is
> doing the correct thing. I don't like the dummy dependency solution but if
> that's what Yaqi wants to do, that's fine.
>
> On Thu, Mar 2, 2017 at 12:16 AM Alex Lindsay <notifications@github.com>
> wrote:
>
> > It's for me to tell without seeing the Rattlesnake code, but is this
> > truly some unorthodox calculation case? Because if the PR isn't allowing
> > material calculations when it's supposed to be, then the PR isn't doing
> > its job. However, if it's some heavily custom logic in the user object,
> > then perhaps I can forgive it.
> >
> > On 03/01/2017 10:31 PM, Sebastian Schunert wrote:
> > >
> > > As @rwcarlsen <https://github.com/rwcarlsen> mentioned it's the
> > > fission source not being executed on initial that causes the issue.
> > > This appears only for cases where we set the initial condition of a
> > > transient via an eigenvalue.
> > > Adding a dummy pp like this
> > >
> > > |[./a_dummy] type = ElementIntegralMaterialProperty mat_prop =
> > > fission_source execute_on = 'initial' [../] |
> > >
> > > fixes it [trivial diff about a_dummy being added].
> > > We could add a PP in for transients via the ActionSystems @YaqiWang
> > > <https://github.com/YaqiWang>?
> > > As far as I can see, only one test has this issue, the others should
> > > be iterative tolerances.
> > >
> > > —
> > > You are receiving this because you were mentioned.
> > > Reply to this email directly, view it on GitHub
> > > <https://github.com/idaholab/moose/pull/8467#issuecomment-283553372>,
> > > or mute the thread
> > > <
> > https://github.com/notifications/unsubscribe-auth/
> AJxgcNQByC7GCqQCrGzaFz5QYPOZ-JQuks5rhkYIgaJpZM4L1wRM
> > >.
> > >
> >
> > —
> > You are receiving this because you were mentioned.
> >
> >
> > Reply to this email directly, view it on GitHub
> > <https://github.com/idaholab/moose/pull/8467#issuecomment-283558801>,
> or mute
> > the thread
> > <https://github.com/notifications/unsubscribe-auth/
> AC5XIMhcsPyL2K5hN3RXvLEoNHIodro6ks5rhlCmgaJpZM4L1wRM>
> > .
> >
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/8467#issuecomment-283630355>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AHLQEwHl9SqSpSQknzCHaJ_V9YJ4Znxwks5rhqmQgaJpZM4L1wRM>
> .
>

rwcarlsen commented:
@snschune - You should be able to see my rattlesnake repo now.

rwcarlsen commented:
I instrumented MOOSE (on top of this PR) and watched the value of some BISON material properties change from an extra material property evaluation inserted right after the evaluation when computing aux variables.  Instrumented moose below:
```diff
diff --git a/framework/src/base/ComputeElemAuxVarsThread.C b/framework/src/base/ComputeElemAuxVarsThread.C
index 170677f..a0d6f4c 100644
--- a/framework/src/base/ComputeElemAuxVarsThread.C
+++ b/framework/src/base/ComputeElemAuxVarsThread.C
@@ -77,6 +77,7 @@ ComputeElemAuxVarsThread::subdomainChanged()
 void
 ComputeElemAuxVarsThread::onElement(const Elem * elem)
 {
+  std::cout << "computing elem aux vars thread\n";
   if (_aux_kernels.hasActiveBlockObjects(_subdomain, _tid))
   {
     const std::vector<std::shared_ptr<AuxKernel>> & kernels = _aux_kernels.getActiveBlockObjects(_subdomain, _tid);
@@ -103,6 +104,10 @@ ComputeElemAuxVarsThread::onElement(const Elem * elem)
       }
     }
   }
+  std::cout << "do-extra-material-evaluation\n";
+  _fe_problem.reinitMaterials(elem->subdomain_id(), _tid);
+
+  std::cout << "done computing elem aux vars thread\n";
 }

 void
diff --git a/framework/src/base/FEProblemBase.C b/framework/src/base/FEProblemBase.C
index 0a0a583..08b8b26 100644
--- a/framework/src/base/FEProblemBase.C
+++ b/framework/src/base/FEProblemBase.C
@@ -2047,6 +2047,11 @@ FEProblemBase::reinitMaterials(SubdomainID blk_id, THREAD_ID tid, bool swap_stat

     if (_materials.hasActiveBlockObjects(blk_id, tid))
       _material_data[tid]->reinit(_materials.getActiveBlockObjects(blk_id, tid));
+
+    for (auto& mat : _materials.getActiveBlockObjects(blk_id, tid))
+      if (mat->hasMaterialProperty<Real>("gas_swell"))
+        for (int i = 0; i < n_points; i++)
+          std::cout << "   gas_swell[" << i << "]=" << mat->getMaterialProperty<Real>("gas_swell")[i] << "\n";
   }
 }
```

The ``gas_swell`` property changes values with the extra evaluation when it should be basically idempotent.  This extra call was previously perturbing results, and this PR removes the extra unnecessary computation.  I don't understand the failing test well, but it seems to me that BISON test results might have been either wrong before or dangerously relying on MOOSE implementation details.

YaqiWang commented:
Sorry @snschune I just noticed your comments. I do not like the dummy fix. Typically we have delayed neutron precursors which depend on fission source material property because we assume initial equilibrium condition. Can you let me know which test is failing?

snschune commented:
The PN transient test started from an eigenvalue without DNP if I remember
correctly...So in this case no DNP and fission source is not evaluated on
initial. However, what triggers evaluation of the DNPs if they are not used
on initial?

On Mar 7, 2017 11:49 AM, "Yaqi" <notifications@github.com> wrote:

> Sorry @snschune <https://github.com/snschune> I just noticed your
> comments. I do not like the dummy fix. Typically we have delayed neutron
> precursors which depend on fission source material property because we
> assume initial equilibrium condition. Can you let me know which test is
> failing?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/8467#issuecomment-284818427>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AHLQE2_uPZZvgS9i4gnwc8pHTAaj2RE8ks5rjabHgaJpZM4L1wRM>
> .
>

YaqiWang commented:
No idea. Can you push what you have in Rattlesnake and let me take a look?

snschune commented:
There is no code I've added wrt to this. The dummy evaluation via action
system was a suggestion but I haven't implemented it. If you want to take a
look pull this branch for moose and Robert's branch for rattlesnake. You
can make the test pass by simply adding the dummy evaluation in the input
file.

On Mar 7, 2017 12:20 PM, "Yaqi" <notifications@github.com> wrote:

> No idea. Can you push what you have in Rattlesnake and let me take a look?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/8467#issuecomment-284827241>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AHLQEw8zsTObE6P_tSPE4SOQOOPL0Z6vks5rja4ZgaJpZM4L1wRM>
> .
>

rwcarlsen commented:
@bwspenc - Can BISON folks look into the BISON failures and confirm/refute my suspicions noted above (i.e. https://github.com/idaholab/moose/pull/8467#issuecomment-284806330)?  I think either those BISON tests will need to be regolded or a BISON bug fix is in order.

bwspenc commented:
@rwcarlsen sure we can look into it.  I'm in a meeting today, but maybe someone else can look into it.  @sapitts @gambka ?

gambka commented:
@rwcarlsen I'll take a look at the BISON failures.

gambka commented:
@rwcarlsen I've taken a look at the BISON failures. You mentioned in a previous comment that you observed the `gas_swell` property change when an extra material property evaluation was inserted right after the evaluation when computing aux variables.  By looking at your instrumented MOOSE this extra material property evaluation occurs after the computation of Elemental Aux Variables.  I believe the reason you notice `gas_swell` change is because it depends upon the variable hydrostatic_stress.  This variable is a CONSTANT MONOMIAL ElementalAux with execute_on timestep_end in the test.

It would seem to me that the extra material evaluation is using the latest computed value for the aux variable which is causing the material property to change.

rwcarlsen commented:
In summary - BISON has a material that depends on an aux variable.  Materials are computed *before* aux variables.  Previously, MOOSE would re-evaluate all materials for all loops (e.g. aux var, residual, user objects etc.).  In BISON, there are 2 tests where material output data are recorded.  Previous to this Pr, those materials were being evaluated, aux variables computed, and then those variables re-evaluated all on a single time step.  So prior to this PR, BISON was (sort of) recording/outputting the "next" steps data in the current step due to extra post-aux var calc material re-evaluations occurring in the compute user objects thread.  With this PR, materials are only evaluated when needed - not getting the extra evaluation after aux var computation it was before (via user objects calculation).  Does that seem correct?

moosebuild commented:
Job `Test intel:linux-intel` on 188efec : invalidated by @rwcarlsen

rwcarlsen commented:
Have PRs into mastodon and rattlesnake that will fix their errors. Just waiting on a verdict from BISON team.

gambka commented:
@rwcarlsen That description seems correct to me.  There is an improved way to handle the hydrostatic stress for those tests that would require some changes to the code.  However, based upon the changes in this PR and how the code and tests are currently constructed in BISON I believe it would be sufficient to regold these tests such that this PR can be merged.  What do you think @bwspenc ?

moosebuild commented:
Job `App tests:linux-gnu` on 188efec : invalidated by @rwcarlsen

rwcarlsen commented:
ping - ready.

permcody commented:
Since this will cause failure, have one of the guys in the cube coordinate the merge with you. I don't like to click the button on these types of merges when I'm out of the office.

lindsayad commented:
Thanks @rwcarlsen for helping this get through!

YaqiWang commented:
The fix in Rattlesnake is probably not very right, it is currently causing a test `treat_5x5` randomly failing. What I can get is
```
>>>> EmbeddedMeshVolumeVariableTransfer is executed.
/apps/local/easybuild/software/GCC/4.9.2/include/c++/4.9.2/debug/vector:357:
    error: attempt to subscript container with out-of-bounds index 0, but 
    container only holds 0 elements.

Objects involved in the operation:
sequence "this" @ 0x0x137284d8 {
  type = NSt7__debug6vectorIP16MaterialPropertyIdESaIS3_EEE;
}
[falcon1:mpi_rank_8][error_sighandler] Caught error: Aborted (signal 6)
```
suggest something wrong. I will suggest @rwcarlsen  working with @snschune, who can explain what the code is doing there, to fix it.

rwcarlsen commented:
@YaqiWang - that looks eerily similar to the other crashing problems that I (supposedly) fixed in rattlesnake.  I suspect it has to do with rattlesnake accessing material properties **not** via the material property interface at a time when they are not necessarily initialized yet depending on the particular order MOOSE iterates over objects to initialize early in the simulation.  Let's move this to a rattlesnake issue and continue further investigation there.

YaqiWang commented:
@rwcarlsen We can use this https://hpcgitlab.inl.gov/idaholab/rattlesnake/issues/578 for fixing the issue.
